### ISSUE TITLE
[no-invalid-void-type] warns `void` at default value of optional type parameter

### ISSUE THREAD

**Issue by ryym** [typescript-eslint/typescript-eslint#4695]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-invalid-void-type": [
      "error",
      { "allowInGenericTypeArguments": true }
    ]
  }
}
```

```TS
// index.ts
type Func<R = void> = () => R;
```

**Expected Result**

ESLint reports no errors since `allowInGenericTypeArguments` is true.

**Actual Result**

ESLint reports an error.

```
npx eslint index.ts
/Users/yamada.ryutaro/work/ts-eslint-void/index.ts
  1:15  error  void is only valid as a return type or generic type variable  @typescript-eslint/no-invalid-void-type

✖ 1 problem (1 error, 0 warnings)
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.15.0` |
| `@typescript-eslint/parser`        | `5.15.0` |
| `TypeScript`                       | `4.6.2` |
| `ESLint`                           | `8.11.0` |
| `node`                             | `16.14.1` |

**Comment by bradzacher** [comment]
This is actually not a false positive!
```ts
type Func<R = void> = () => R;
//        ^^^^^^^^ type **parameter** declaration
```

An argument is the name of a thing passed as a call.
A parameter is the name of the variables a call accepts.
Or another way:

```ts
function foo(one, two, three) { ... }
//           ^^^^^^^^^^^^^^^ parameters

type Foo<T1, T2, T3> = ...;
//       ^^^^^^^^^^ parameters

foo(one, two, three);
//  ^^^^^^^^^^^^^^^ arguments

type Bar = Foo<T1, T2, T3>;
//             ^^^^^^^^^^ arguments
```

This is an intentional distinction as it's very common for people to consume generic types which use the generic in a function return location, but it's uncommon for people to create generic types which place the generic in a function return location AND specify a default value for the generic.

In this case I would suggest using a disable comment.

**Comment by ryym** [comment]
Thanks, I understood.
I think it would be a pain to have to use a disable comment for code that isn't problematic, so I am wondering if there should be a new option to allow `void` in type parameter declarations. However, it's not a big problem for me at the moment, so I'll open a new issue or PR in the future when this issue is upvoted.

- `allowInGenericTypeDeclaration: boolean`
- allow `T = void` , `T extends U ? void : void`
- disallow `T extends void` since it doesn't make sense
